### ISSUE TITLE
Double / Long deserialize

### ISSUE THREAD

**Issue by yuvalp-k2view** [google/gson#1679]
Sometimes it is useful to parse '5' as long vs '5.0' as double, especially when parsing arbitrary data.

This would make the serialize/deserialize of numbers symetric (5.0 > "5.0" > 5.0, 5L > "5" > 5L)

I'd like to suggest a new builder param:
setNumberDererializationPolicy()
default will remain as is
however, user will be able to specify:
setNumberDererializationPolicy(NumberDererializationPolicy.LONG_SUPPORT)

I think this presents a good compromise between the spec being strictly double and the reality where double has precision issues and users need a way to store longs in arbitrary json.

We'd be happy to create a PR if this is acceptable

**Comment by lyubomyr-shaydariv** [comment]
Possible duplicate of https://github.com/google/gson/pull/1290 :

```java
Gson gson = new GsonBuilder()
  .setNumberToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
  .create();
```

**Comment by yuvalp-k2view** [comment]
I agree... but I see this PR is waiting for a long time. Do you know if there is anything wrong with it? If there are fixes needed we'd be happy to contribute.

**Comment by lyubomyr-shaydariv** [comment]
@yuvalp-k2view 

> Do you know if there is anything wrong with it?

No idea, there was no review from the project members.
